File: HWs/TimeSeriesDeseasonalization/scripts/julia/deseasonalization.jl
using DataFrames, CSV, Statistics, GLM, Plots, StatsBase, StatsPlots, Distributions

df = CSV.read("../../data/avocado_exports.csv", DataFrame)
df.Año = Int.(df.Año)
df.Bimestre = Int.(df.Bimestre)
df.t = Int.(df.t)
df.Toneladas = Float64.(df.Toneladas)

media = mean(df.Toneladas)
desv = std(df.Toneladas)
cv = (desv / media) * 100
println("[1] Coeficiente de variación: $(round(cv, digits=2))%")

println("\n[2] Estadísticas descriptivas (Toneladas):")
println(describe(df.Toneladas))

p1 = plot(df.t, df.Toneladas, marker=:circle, 
          title="Serie bimestral de exportaciones de aguacate",
          xlabel="t (bimestres desde 2019-1)", ylabel="Toneladas",
          grid=true, legend=false)
savefig(p1, "../../plots/julia/avocado_exports_bimestral.png")
display(p1)

p2 = histogram(df.Toneladas, bins=10,
               title="Distribución de exportaciones (Toneladas)",
               xlabel="Toneladas", ylabel="Frecuencia",
               grid=true, legend=false)
savefig(p2, "../../plots/julia/avocado_exports_histogram.png")
display(p2)

p3 = boxplot(df.Toneladas, title="Boxplot de exportaciones bimestrales (Toneladas)",
              ylabel="Toneladas", grid=true, legend=false)
savefig(p3, "../../plots/julia/avocado_exports_boxplot.png")
display(p3)

X = [ones(size(df.t, 1)) df.t]
y = df.Toneladas
model_orig = lm(X, y)

pendiente = coef(model_orig)[2] / 1000.0
println("\n[3] Coeficiente de pendiente muestral (miles de toneladas por bimestre): $(round(pendiente, digits=4))")

println("\n[4] Prueba de hipótesis para pendiente > 0 (serie original)")
println(model_orig)

t_stat = coef(model_orig)[2] / stderror(model_orig)[2]
p_two_sided = 2 * (1 - cdf(TDist(dof_residual(model_orig)), abs(t_stat)))
p_one_sided = p_two_sided / 2
if t_stat < 0
    p_one_sided = 1 - p_one_sided
end

println("t = $(round(t_stat, digits=3)), p una-cola = $(round(p_one_sided, digits=4)) -> $(p_one_sided < 0.05 ? "Rechazamos H0: tendencia positiva" : "No se rechaza H0")")

k = 6
promedio_global = mean(df.Toneladas)
indices_estacionales = Dict()
for bim in 1:k
    indices_estacionales[bim] = mean(df.Toneladas[df.Bimestre .== bim]) / promedio_global
end

println("\n[5] Índices estacionales (multiplicativos):")
for bim in 1:k
    println("Bimestre $bim: $(round(indices_estacionales[bim], digits=4))")
end
println("Interpretación: valores >1 indican bimestres por encima del promedio; <1 por debajo.")

df.IndiceEstacional = [indices_estacionales[bim] for bim in df.Bimestre]
df.Deseasonalizada = df.Toneladas ./ df.IndiceEstacional

p4 = plot(df.t, df.Deseasonalizada, marker=:circle,
          title="Serie desestacionalizada (multiplicativa)",
          xlabel="t", ylabel="Toneladas desestacionalizadas",
          grid=true, legend=false)
savefig(p4, "../../plots/julia/avocado_exports_deseasonalized.png")
display(p4)

X_d = [ones(size(df.t, 1)) df.t]
y_d = df.Deseasonalizada
model_des = lm(X_d, y_d)

println("\n[6] Regresión con datos desestacionalizados:")
println(model_des)

t_stat_d = coef(model_des)[2] / stderror(model_des)[2]
p_two_sided_d = 2 * (1 - cdf(TDist(dof_residual(model_des)), abs(t_stat_d)))
p_one_sided_d = p_two_sided_d / 2
if t_stat_d < 0
    p_one_sided_d = 1 - p_one_sided_d
end

println("t = $(round(t_stat_d, digits=3)), p una-cola = $(round(p_one_sided_d, digits=4)) -> $(p_one_sided_d < 0.05 ? "Rechazamos H0: tendencia positiva" : "No se rechaza H0")")

t_future = 32
X_future = [1.0 t_future]
pred_mean = predict(model_orig, X_future)[1]
V = vcov(model_orig)
x0 = [1.0, t_future]
se_pred = sqrt(x0' * V * x0)
t_critical = quantile(TDist(dof_residual(model_orig)), 0.975)
ci_low = pred_mean - t_critical * se_pred
ci_high = pred_mean + t_critical * se_pred

println("\n[7] IC 95% para la tendencia (media esperada) en 2024-bim2 (t=32): [$(round(ci_low, digits=2)), $(round(ci_high, digits=2))] toneladas. Estimado: $(round(pred_mean, digits=2))")

r2_orig = r2(model_orig)
r2_des = r2(model_des)
println("\n[8] R^2 serie original: $(round(r2_orig, digits=4))")
println("[8] R^2 serie desestacionalizada: $(round(r2_des, digits=4))")
println("Explicación: al quitar la variación estacional, el componente sistemático por tiempo puede capturar mejor la tendencia subyacente (o a veces menos, si la estacionalidad ya explicaba variación alineada con t). Un cambio en R^2 refleja cuánto peso de la variabilidad se atribuye a la estacionalidad vs. la tendencia.")

fut = DataFrame(
    t = 31:36
)

function t_to_year_bim(t)
    base_year = 2019
    idx = t - 1
    year = base_year + div(idx, 6)
    bim = mod(idx, 6) + 1
    return year, bim
end

yb = [t_to_year_bim(t) for t in fut.t]
fut.Año = [y[1] for y in yb]
fut.Bimestre = [y[2] for y in yb]

X_fut = [ones(size(fut.t, 1)) fut.t]
pred_orig = predict(model_orig, X_fut)
fut.Pronostico_Original = pred_orig

pred_des = predict(model_des, X_fut)
fut.IndiceEstacional = [indices_estacionales[bim] for bim in fut.Bimestre]
fut.Pronostico_Deseason = pred_des .* fut.IndiceEstacional

println("\n[9] Pronósticos 2024 (toneladas):")
println(round.(fut[:, [:Año, :Bimestre, :t, :Pronostico_Original, :Pronostico_Deseason]], digits=2))

p5 = plot(df.t, df.Toneladas, marker=:circle, label="Observado")
plot!(fut.t, fut.Pronostico_Original, marker=:circle, linestyle=:dash, label="Pronóstico (original)")
title!("Pronóstico con modelo original")
xlabel!("t")
ylabel!("Toneladas")
plot!(grid=true)
savefig(p5, "../../plots/julia/avocado_exports_original_forecast.png")
display(p5)

p6 = plot(df.t, df.Toneladas, marker=:circle, label="Observado")
plot!(fut.t, fut.Pronostico_Deseason, marker=:circle, linestyle=:dash, label="Pronóstico (desestacionalizado)")
title!("Pronóstico con modelo desestacionalizado (reestacionalizado)")
xlabel!("t")
ylabel!("Toneladas")
plot!(grid=true)
savefig(p6, "../../plots/julia/avocado_exports_deseasonalized_forecast.png")
display(p6)

File: HWs/TimeSeriesDeseasonalization/scripts/python/deseasonalization.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm

df = pd.read_csv("../../data/avocado_exports.csv")
df["Año"] = df["Año"].astype(int)
df["Bimestre"] = df["Bimestre"].astype(int)
df["t"] = df["t"].astype(int)
df["Toneladas"] = df["Toneladas"].astype(float)

media = df["Toneladas"].mean()
desv = df["Toneladas"].std(ddof=1)
cv = (desv / media) * 100
print(f"[1] Coeficiente de variación: {cv:.2f}%")

print("\n[2] Estadísticas descriptivas (Toneladas):")
print(df["Toneladas"].describe())

plt.figure()
plt.plot(df["t"], df["Toneladas"], marker="o")
plt.title("Serie bimestral de exportaciones de aguacate")
plt.xlabel("t (bimestres desde 2019-1)")
plt.ylabel("Toneladas")
plt.grid(True)
plt.savefig("../../plots/python/avocado_exports_bimestral.png")
plt.show()

plt.figure()
plt.hist(df["Toneladas"], bins=10)
plt.title("Distribución de exportaciones (Toneladas)")
plt.xlabel("Toneladas")
plt.ylabel("Frecuencia")
plt.grid(True)
plt.savefig("../../plots/python/avocado_exports_histogram.png")
plt.show()

plt.figure()
plt.boxplot(df["Toneladas"], vert=True, showmeans=True)
plt.title("Boxplot de exportaciones bimestrales (Toneladas)")
plt.ylabel("Toneladas")
plt.grid(True)
plt.savefig("../../plots/python/avocado_exports_boxplot.png")
plt.show()

X = sm.add_constant(df["t"])
y = df["Toneladas"]
model_orig = sm.OLS(y, X).fit()

pendiente = model_orig.params["t"] / 1000.0
print(
    f"\n[3] Coeficiente de pendiente muestral (miles de toneladas por bimestre): {pendiente:.4f}"
)

print("\n[4] Prueba de hipótesis para pendiente > 0 (serie original)")
print(model_orig.summary())

t_stat = model_orig.tvalues["t"]
p_two_sided = model_orig.pvalues["t"]
p_one_sided = p_two_sided / 2 if t_stat > 0 else 1 - p_two_sided / 2
print(
    f"t = {t_stat:.3f}, p una-cola = {p_one_sided:.4g}  -> "
    f"{'Rechazamos H0: tendencia positiva' if p_one_sided < 0.05 else 'No se rechaza H0'}"
)

k = 6
promedio_global = df["Toneladas"].mean()
indices_estacionales = df.groupby("Bimestre")["Toneladas"].mean() / promedio_global
indices_estacionales = indices_estacionales.reindex(range(1, k + 1))

print("\n[5] Índices estacionales (multiplicativos):")
for bim, ind in indices_estacionales.items():
    print(f"Bimestre {bim}: {ind:.4f}")
print(
    "Interpretación: valores >1 indican bimestres por encima del promedio; <1 por debajo."
)

df["IndiceEstacional"] = df["Bimestre"].map(indices_estacionales)
df["Deseasonalizada"] = df["Toneladas"] / df["IndiceEstacional"]

plt.figure()
plt.plot(df["t"], df["Deseasonalizada"], marker="o")
plt.title("Serie desestacionalizada (multiplicativa)")
plt.xlabel("t")
plt.ylabel("Toneladas desestacionalizadas")
plt.grid(True)
plt.savefig("../../plots/python/avocado_exports_deseasonalized.png")
plt.show()

X_d = sm.add_constant(df["t"])
y_d = df["Deseasonalizada"]
model_des = sm.OLS(y_d, X_d).fit()
print("\n[6] Regresión con datos desestacionalizados:")
print(model_des.summary())

t_stat_d = model_des.tvalues["t"]
p_two_sided_d = model_des.pvalues["t"]
p_one_sided_d = p_two_sided_d / 2 if t_stat_d > 0 else 1 - p_two_sided_d / 2
print(
    f"t = {t_stat_d:.3f}, p una-cola = {p_one_sided_d:.4g}  -> "
    f"{'Rechazamos H0: tendencia positiva' if p_one_sided_d < 0.05 else 'No se rechaza H0'}"
)

t_future = 32
X_future = pd.DataFrame({"const": 1.0, "t": [t_future]})
pred_mean = model_orig.get_prediction(X_future)
ci_low, ci_high = pred_mean.conf_int(alpha=0.05)[0]
mean_pred = pred_mean.predicted_mean[0]
print(
    f"\n[7] IC 95% para la tendencia (media esperada) en 2024-bim2 (t=32): "
    f"[{ci_low:.2f}, {ci_high:.2f}] toneladas. Estimado: {mean_pred:.2f}"
)
pi = pred_mean.summary_frame(alpha=0.05)[["obs_ci_lower", "obs_ci_upper"]].values[0]
print(
    f"[7-extra] IC 95% de predicción (valor observado esperado) en 2024-bim2: [{pi[0]:.2f}, {pi[1]:.2f}]"
)

r2_orig = model_orig.rsquared
r2_des = model_des.rsquared
print(f"\n[8] R^2 serie original: {r2_orig:.4f}")
print(f"[8] R^2 serie desestacionalizada: {r2_des:.4f}")
print(
    "Explicación: al quitar la variación estacional, el componente sistemático por tiempo "
    "puede capturar mejor la tendencia subyacente (o a veces menos, si la estacionalidad "
    "ya explicaba variación alineada con t). Un cambio en R^2 refleja cuánto peso de la "
    "variabilidad se atribuye a la estacionalidad vs. la tendencia."
)

fut = pd.DataFrame(
    {
        "t": np.arange(31, 37),
    }
)


def t_to_year_bim(t):
    base_year = 2019
    idx = t - 1
    year = base_year + idx // 6
    bim = (idx % 6) + 1
    return year, bim


yb = np.array([t_to_year_bim(t) for t in fut["t"]])
fut["Año"] = yb[:, 0]
fut["Bimestre"] = yb[:, 1]
fut["const"] = 1.0

pred_orig = model_orig.get_prediction(fut[["const", "t"]])
fut["Pronostico_Original"] = pred_orig.predicted_mean

pred_des = model_des.get_prediction(fut[["const", "t"]]).predicted_mean
fut["IndiceEstacional"] = fut["Bimestre"].map(indices_estacionales)
fut["Pronostico_Deseason"] = pred_des * fut["IndiceEstacional"]

print("\n[9] Pronósticos 2024 (toneladas):")
print(
    fut[["Año", "Bimestre", "t", "Pronostico_Original", "Pronostico_Deseason"]].round(2)
)

pred_orig_full = model_orig.get_prediction(fut[["const", "t"]])
pred_des_full = model_des.get_prediction(fut[["const", "t"]])

fut["IC_orig_low"], fut["IC_orig_high"] = pred_orig_full.conf_int(alpha=0.05).T
pi_des = pred_des_full.summary_frame(alpha=0.05)
fut["IC_des_low"], fut["IC_des_high"] = (
    pi_des["obs_ci_lower"] * fut["IndiceEstacional"],
    pi_des["obs_ci_upper"] * fut["IndiceEstacional"],
)

print("\n[9-extra] Pronósticos con intervalos de confianza:")
print(
    fut[
        [
            "Año",
            "Bimestre",
            "Pronostico_Original",
            "IC_orig_low",
            "IC_orig_high",
            "Pronostico_Deseason",
            "IC_des_low",
            "IC_des_high",
        ]
    ].round(2)
)

plt.figure()
plt.plot(df["t"], df["Toneladas"], marker="o", label="Observado")
plt.plot(
    fut["t"],
    fut["Pronostico_Original"],
    marker="o",
    linestyle="--",
    label="Pronóstico (original)",
)
plt.title("Pronóstico con modelo original")
plt.xlabel("t")
plt.ylabel("Toneladas")
plt.legend()
plt.grid(True)
plt.savefig("../../plots/python/avocado_exports_original_forecast.png")
plt.show()

plt.figure()
plt.plot(df["t"], df["Toneladas"], marker="o", label="Observado")
plt.plot(
    fut["t"],
    fut["Pronostico_Deseason"],
    marker="o",
    linestyle="--",
    label="Pronóstico (desestacionalizado)",
)
plt.title("Pronóstico con modelo desestacionalizado (reestacionalizado)")
plt.xlabel("t")
plt.ylabel("Toneladas")
plt.legend()
plt.grid(True)
plt.savefig("../../plots/python/avocado_exports_deseasonalized_forecast.png")
plt.show()
